README.md
# fastify-webhook

  [![NPM Version](https://img.shields.io/npm/v/fastify-webhook.svg?style=flat)](https://npmjs.org/package/fastify-webhook/)
  [![NPM Downloads](https://img.shields.io/npm/dm/fastify-webhook.svg?style=flat)](https://npmjs.org/package/fastify-webhook/)
  [![Code Style](https://img.shields.io/badge/code%20style-standard-brightgreen.svg?style=flat)](http://standardjs.com/)
  [![license - APACHE-2.0](https://img.shields.io/npm/l/fastify-webhook.svg)](http://opensource.org/licenses/APACHE-2.0)

Fastify Plugin to serve webhooks with some useful default settings.

With this plugin, Fastify will have a route configured for `/webhook` POST requests.


## Usage

```js
const fastify = require('fastify')()

// example without specifying options, returning a default webhook mapped to '/webhook' that only acknowledge the POST request
fastify.register(require('fastify-webhook'))
// or
// example with custom webhook url and handler, and secret key
// fastify.register(require('fastify-webhook'), {'url': '/custom-webhook', 'handler': myWebhookHandler, 'secretKey': 'secret key'})
//
// note that to use one of handlers bundled with the plugin, you need to get a reference to the plugin script 'src/handlers', and then as handler pass a reference to desired function, like:
// const webhookHandlers = require('fastify-webhook/src/handlers') // get plugin handlers (optional)
// const webhookPlugin = require('fastify-webhook')
// fastify.register(webhookPlugin, { 'url': '/custom-webhook', 'handler': webhookHandlers.echo, 'secretKey': 'secret key'})
//

fastify.listen(3000)

// To test, for example (in another terminal session) do:
// curl http://127.0.0.1:3000/webhook -X POST -H 'Content-Type: application/json' -d '{"payload":"test"}' => returning a JSON dump of the given data, and no thrown error
// in Windows you need to escape double quote char in the given json body, so do:
// curl http://127.0.0.1:3000/webhook -X POST -H "Content-Type: application/json" -d "{\"payload\":\"test\"}"
// or put data in a json file and pass with something like: '-d @body.json'
```

In the [example](./example/) folder there are some simple server scripts that uses the plugin (inline but it's the same using it from npm registry).


## Requirements

Fastify ^4.0.1 , Node.js 14 LTS (14.15.0) or later.
Note that plugin releases 3.x are for Fastify 3.x, 4.x are for Fastify 4.x, etc.


## Sources

Source code is all inside main repo:
[fastify-webhook](https://github.com/smartiniOnGitHub/fastify-webhook).

Documentation generated from source code (library API):
[here](https://smartiniongithub.github.io/fastify-webhook/).


## Note

By default the plugin map a default handler on the URI `/webhook` to be called via POST, otherwise it's possible to change via the setting 'url' in plugin options.

The plugin exposes some handlers, for common base operations (and webhook debug help).
To use one of them, before registering the plugin, you need to get a reference from its `src/handlers` source file;
then you can configure the desired one in the setting 'handler' in plugin options.
They are:
- `acknowledge` (default handler) that simply acknowledge the request, and reply with a simple json response
- `echo` it dumps the given input data in the (json) response
- `logger` it dumps some info on the request using Fastify logger
but of course for a real world usage you need to specify your own handler function, with arguments '(request, reply)'.
Otherwise you can use yours, with signature `function handler (request, reply)`.

Other plugin options:
- 'disableWebhook' (default false) to disable the registration of the route for the webhook
- 'enableGetPlaceholder' (default false) to publish via GET a route with the same path of the webhook but returning an HTTP Error 405 Method Not Allowed
- 'secretKey' (default null) to specify a string as secret key that callers of the webhook must provide, or webhook will reply with an error
- 'preHandlers' is a list of functions to be used as preHandler in the specific route of the webhook; currently the list contains an internal function to check the secret key (if given); otherwise you can define and use yours, with signature `function preHandler (request, reply, done)`.

Note that there is not a good general way to handle (usually user-specific) token in requests, so this is not managed via the plugin, but in examples and tests you can find some info.


## License

Licensed under [Apache-2.0](./LICENSE).

----


CHANGELOG.md
# Change Log

## [4.0.1](https://github.com/smartiniOnGitHub/fastify-webhook/releases/tag/4.0.1) (2022-07-25)
[Full Changelog](https://github.com/smartiniOnGitHub/fastify-webhook/compare/4.0.0...4.0.1)
Summary Changelog:
- Removed reference to an internal Fastify data structure for the request

## [4.0.0](https://github.com/smartiniOnGitHub/fastify-webhook/releases/tag/4.0.0) (2022-07-22)
[Full Changelog](https://github.com/smartiniOnGitHub/fastify-webhook/compare/3.0.0...4.0.0)
Summary Changelog:
- Updated requirements to Fastify '^4.0.0' and Fastify-plugin '^4.0.0'; 
  update code with related changes
- Updated all dependencies to latest (for Node.js 14 LTS)
- Update and simplified example and test code
- Update documentation from sources with JSDoc

## [3.0.0](https://github.com/smartiniOnGitHub/fastify-webhook/releases/tag/3.0.0) (2022-03-27)
[Full Changelog](https://github.com/smartiniOnGitHub/fastify-webhook/compare/2.1.0...3.0.0)
Summary Changelog:
- Update requirements to latest Fastify 3.x and Node.js 10 LTS
- Ensure all works again

## [2.1.0](https://github.com/smartiniOnGitHub/fastify-webhook/releases/tag/2.1.0) (2021-04-07)
[Full Changelog](https://github.com/smartiniOnGitHub/fastify-webhook/compare/2.0.0...2.1.0)
Summary Changelog:
- Update requirements to latest Fastify 2.x, so currently release '^2.15.3'
- Feature: keep compatibility with Node.js 8 LTS (but this is last release)
- Update all other dependencies
- Update Tap to latest (major), and update deprecated calls in tests
- Add JSDoc to generate documentation from sources

## [2.0.0](https://github.com/smartiniOnGitHub/fastify-webhook/releases/tag/2.0.0) (2019-04-10)
[Full Changelog](https://github.com/smartiniOnGitHub/fastify-webhook/compare/1.0.0...2.0.0)
Summary Changelog:
- Update requirements to Fastify v2
- Update all dependencies
- Breaking Change: rename plugin option 'beforeHandler' to 'preHandler', 
  due to a change/rename in Fastify v2
- Update function arguments like in Fastify v2 (from req to request, etc)

## [1.0.0](https://github.com/smartiniOnGitHub/fastify-webhook/releases/tag/1.0.0) (2019-04-05)
Summary Changelog:
- Updated all dependencies
- Note that this release number means that the plugin is stable, 
  and for Fastify v1
- Small updates

## [0.4.0](https://github.com/smartiniOnGitHub/fastify-webhook/releases/tag/0.4.0) (2019-04-01)
Summary Changelog:
- Align with latest Fastify v1 (and related plugins)
- Updated all dependencies
- Updated Tap tests
- Update code to use latest ES6/7/8 syntax
- Improve tests and examples, to read 'secretKey' value from env var SECRET_KEY 
  or otherwise from a fixed default value
- Breaking Change: rename plugin option 'disableDefaultWebhook' into 'disableWebhook'
- Add plugin option 'enableGetPlaceholder' (default false) to publish via GET 
  a route with the same path of the webhook, 
  but returning an HTTP Error 405 Method Not Allowed;
  add a sample usage in the 'example-enhanced'

## [0.3.2](https://github.com/smartiniOnGitHub/fastify-webhook/releases/tag/0.3.2) (2019-02-03)
Summary Changelog:
- Updated all dependencies
- Updated Tap tests

## [0.3.1](https://github.com/smartiniOnGitHub/fastify-webhook/releases/tag/0.3.1) (2019-01-02)
Summary Changelog:
- Update dependency on 'fastify-plugin' to '^1.4.0'
- Updated all other plugin dependencies

## [0.3.0](https://github.com/smartiniOnGitHub/fastify-webhook/releases/tag/0.3.0) (2018-12-03)
Summary Changelog:
- Update required Fastify version to '^1.1.0', but stay on 1.x
- Update dependency on 'fastify-plugin' to '^1.2.1'
- Updated plugins
- Updated unit tests and examples to use new Fastify syntax/features
- In examples, listen to '127.0.0.1' and no more to all addresses ('0.0.0.0'),
  best practice for examples (unless you need to publish from a container)
- Move plugin main source in a 'src' folder
- Small cleanup and simplifications in the code base

## [0.2.0](https://github.com/smartiniOnGitHub/fastify-webhook/releases/tag/0.2.0) (2018-05-29)
Summary Changelog:
- Changed dev dependencies from 'request' to 'simple-get' and related tests, like in Fastify and related core plugins
- Add an example using secret key and user token, to run with npm custom command 'example-enhanced'
- Simplified a little the 'package.json' file
- Update documentation with some more options

## [0.1.0](https://github.com/smartiniOnGitHub/fastify-webhook/releases/tag/0.1.0) (2018-04-18)
Summary Changelog:
- First release, with basic features: default configuration and the ability to specify custom operations
- Provide some basic handlers (acknowledge, echo, logger)
- Optional secret key that consumers must provide to call the webhook exposed

----


src/plugin.js
/*
 * Copyright 2018-2022 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
'use strict'

const fp = require('fastify-plugin')

const webhookHandlers = require('./handlers') // get plugin handlers for default handler

function fastifyWebHook (fastify, options, next) {
  const {
    url = '/webhook',
    handler = webhookHandlers.acknowledge,
    disableWebhook = false,
    enableGetPlaceholder = false,
    secretKey = null,
    preHandlers = [checkSecretKey]
  } = options

  function checkSecretKey (request, reply, done) {
    if (secretKey) {
      const contentType = request.headers['content-type'] || ''
      const requestSecretKey = (request.body) ? request.body.secretKey : ''
      if (request.method !== 'POST' ||
        !contentType.startsWith('application/json') ||
        requestSecretKey !== secretKey) {
        reply.code(403).send(new Error('Missing or wrong secret key'))
      }
    }
    done()
  }

  // check plugin options
  if (typeof url !== 'string') {
    throw new TypeError(`The option url must be a string, instead got a '${typeof url}'`)
  }
  if (typeof handler !== 'function') {
    throw new TypeError(`The option webhook must be a function, instead got a '${typeof handler}'`)
  }
  if (typeof disableWebhook !== 'boolean') {
    throw new TypeError(`The option disableWebhook must be a boolean, instead got a '${typeof disableWebhook}'`)
  }
  if (typeof enableGetPlaceholder !== 'boolean') {
    throw new TypeError(`The option enableGetPlaceholder must be a boolean, instead got a '${typeof enableGetPlaceholder}'`)
  }
  if (secretKey !== null && typeof secretKey !== 'string') {
    throw new TypeError(`The option secretKey must be a string, instead got a '${typeof secretKey}'`)
  }
  if (preHandlers !== null && !Array.isArray(preHandlers)) {
    throw new TypeError(`The option preHandlers must be an array (of functions), instead got a '${typeof preHandlers}'`)
  }

  // execute plugin code
  if (disableWebhook === false) {
    fastify.route({
      method: 'POST',
      url,
      preHandler: preHandlers,
      handler
    })
    if (enableGetPlaceholder === true) {
      fastify.get(url, {}, (request, reply) => {
        reply.code(405).send(new Error('Placeholder for a webhook, you need to call via POST'))
      })
    }
  }

  next()
}

module.exports = fp(fastifyWebHook, {
  fastify: '4.x',
  name: 'fastify-webhook'
})


src/handlers.js
/*
 * Copyright 2018-2022 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
'use strict'

// this module exports some handlers that could be used
// as default and non-default plugin conmfiguration

function _getRequestMimeType (request) {
  return request.headers['content-type']
  // return request.getHeader('content-type')
  // return request.getType()
}

function _defaultSuccessWebhookReply (reply) {
  // return a simple default message for successful processing
  reply.type('application/json').send({ statusCode: 200, result: 'success' })
}

function _failureWebhookReply (reply, message) {
  // return a simple error when processing the request
  reply.type('application/json').send(new Error(message))
}

function _isMimeTypeJSON (req) {
  // tell the MIME Type of the inpout data: true if it's json, false otherwise
  const reqMimeType = _getRequestMimeType(req)
  if (!reqMimeType || reqMimeType !== 'application/json') {
    return false
  }
  return true
}

function acknowledgeWebhookHandler (request, reply) {
  // return a simple acknowledge message
  _defaultSuccessWebhookReply(reply)
}

function loggerWebhookHandler (request, reply) {
  // log the given data with Fastify logger, and return a default acknowledge message
  request.log.info(`Request: MIME Type: "${_getRequestMimeType(request)}", ID: "${request.id}", body: "${request.body}"`)
  _defaultSuccessWebhookReply(reply)
}

function echoWebhookHandler (request, reply) {
  // return a json dump of given input data
  // note that the given data is also logged with Fastify logger
  // but in this case it's important to provide the content type as json, and content data (empty json is not valid) or and error will be raised
  if (!_isMimeTypeJSON(request)) {
    _failureWebhookReply(reply, `Missing or wrong input MIME Type: "${_getRequestMimeType(request)}"`)
    return
  }
  request.log.info(`Request: MIME Type: "${_getRequestMimeType(request)}", ID: "${request.id}", body: "${request.body}"`)
  reply.type('application/json').send(request.body)
}

module.exports = {
  acknowledge: acknowledgeWebhookHandler,
  echo: echoWebhookHandler,
  logger: loggerWebhookHandler
}


example/example.js
/*
 * Copyright 2018-2022 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
'use strict'

const fastify = require('fastify')()

fastify.register(require('../'))

// example to handle a sample home request to serve a static page, optional here
fastify.get('/', function (request, reply) {
  const path = require('path')
  const scriptRelativeFolder = path.join(__dirname, path.sep)
  const fs = require('fs')
  const stream = fs.createReadStream(path.join(scriptRelativeFolder, 'home.html'))
  reply.type('text/html; charset=utf-8').send(stream)
})

fastify.listen({ port: 3000, host: '127.0.0.1' }, (err, address) => {
  if (err) throw err
  console.log(`Server listening on '${address}' ...`)
})

fastify.ready(() => {
  const routes = fastify.printRoutes()
  console.log(`Available Routes:\n${routes}`)

  console.log(`To test the webhook, from another terminal, do something like:
  curl http://127.0.0.1:3000/webhook -X POST -H "Content-Type: application/json" -d '{"payload":"test"}'
  Or similar commands with wget ...
  Note that in Windows curl wants to escape json inner string delimiter.
  `)
})


example/example-with-secretkey-and-token.js
/*
 * Copyright 2018-2022 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
'use strict'

const fastify = require('fastify')({
  // enable logger ...
  logger: {
    level: 'info'
  }
})

// handle secret key (fixed) and user token (user-dependent), as a sample
// if the env var SECRET_KEY is defined, it's value is used, otherwise a default value will be used
const webhookSecretKey = process.env.SECRET_KEY || 'my example Secret Key'

function checkSecretKey (request, reply, done) {
  // function that checks the given secret key, if is good or not
  // note that for simplicity the secret key is mandatory here (not checked in input arguments)
  if (request.headers['content-type'] !== 'application/json' || request.body.secretKey !== webhookSecretKey) {
    reply.code(403).type('application/json').send(new Error('Missing or wrong secret key'))
  }
  done()
}

function checkTokenEven (request, reply, done) {
  const stringToken = request.params.token || ''
  // console.log(`chek token: given "${stringToken}", check if it's even`)
  const numToken = parseInt(stringToken)
  // console.log(`chek token: "${stringToken}" is a number, ${typeof numToken === 'number'}`)
  // console.log(`chek token: "${stringToken}" converted into the number ${numToken}`)
  // function that checks if the given token is good or not
  // for example even numbers are ok, odd or negatives not
  // note that for simplicity the user token is mandatory here (not checked in input arguments)
  if (!isNaN(numToken) && numToken > 0 && (numToken % 2 === 0)) {
    // reply.code(200).type('application/json')
    reply.code(200)
  } else {
    reply.code(403).type('application/json').send(new Error('Missing or wrong token'))
  }
  done()
}

const webhookHandlers = require('../src/handlers') // get plugin handlers (but in a relative way), as a sample
const webhookPlugin = require('../') // get the plugin (but in a relative way), as a sample
fastify.register(webhookPlugin, {
  url: '/custom-webhook/:token',
  handler: webhookHandlers.echo,
  // disableWebhook: false, // same as default
  enableGetPlaceholder: true, // as a sample
  secretKey: webhookSecretKey,
  preHandlers: [checkSecretKey, checkTokenEven]
})

// example to handle a sample home request to serve a static page, optional here
fastify.get('/', function (request, reply) {
  const path = require('path')
  const scriptRelativeFolder = path.join(__dirname, path.sep)
  const fs = require('fs')
  const stream = fs.createReadStream(path.join(scriptRelativeFolder, 'home.html'))
  reply.type('text/html; charset=utf-8').send(stream)
})

fastify.listen({ port: 3000, host: '127.0.0.1' }, (err, address) => {
  if (err) throw err
  console.log(`Server listening on '${address}' ...`)
})

fastify.ready(() => {
  const routes = fastify.printRoutes()
  console.log(`Available Routes:\n${routes}`)

  console.log(`To test the webhook, from another terminal, do something like:
  KO (no secret key, no token): curl http://127.0.0.1:3000/custom-webhook/ -X POST -H "Content-Type: application/json" -d '{"payload":"test"}'
  KO (bad secret key, no token): curl http://127.0.0.1:3000/custom-webhook/ -X POST -H "Content-Type: application/json" -d '{"payload":"test", "secretKey":"my bad Secret Key"}'
  KO (good secret key, bad token): curl http://127.0.0.1:3000/custom-webhook/0999 -X POST -H 'Content-Type: application/json' -d '{"payload":"test", "secretKey":"my example Secret Key"}'
  OK (good secret key, good token): curl http://127.0.0.1:3000/custom-webhook/1000 -X POST -H 'Content-Type: application/json' -d '{"payload":"test", "secretKey":"my example Secret Key"}'
  etc ...
  Or similar commands with wget ...
  Note that in Windows curl wants to escape json inner string delimiter.
  `)
})
